docs/src/interface-definition.md

```diff
@@ -0,0 +1,167 @@
+
+To make implementation easy for everyone, we made it so that you need only to define either allocating or non-allocating force call. The other call is generated is then generated with a macro `AtomsCalculators.@generate_complement`.
+
+Example
+
+```julia
+struct MyType
+end
+
+AtomsCalculators.@generate_complement function AtomsCalculators.forces(system, calculator::Main.MyType; kwargs...)
+    # we can ignore kwargs... or use them to tune the calculation
+    # or give extra information like pairlist
+
+    # add your own definition
+    return zeros(AtomsCalculators.default_force_eltype, length(system))
+end
+```
+
+This creates both `forces` and `forces!`. `AtomsCalculators.default_force_eltype` is a type that can be used to allocate force data. You can also allocate for some other type.
+
+!!! note "Type definition under @generate_complement macro"
+    You need to use explicit definition of type when using
+    `@generate_complement` macro. `Main.MyType` is fine `MyType` is not!
+
+    You also need to define the type before macro call.
+
+Alternatively the definition could have been done with
+
+```julia
+struct MyOtherType
+end
+
+AtomsCalculators.@generate_complement function AtomsCalculators.forces!(f::AbstractVector, system, calculator::Main.MyOtherType; kwargs...)
+    @assert length(f) == length(system)
+    # we can ignore kwargs... or use them to tune the calculation
+    # or give extra information like pairlist
+
+    # add your own definition
+    for i in eachindex(f)
+        f[i] = zero(AtomsCalculators.default_force_eltype)
+    end
+
+    return f
+end
+```
+
+## Other Automatically Generated Calls
+
+Many methods have optimized calls when energy and forces (and/or virial) are calculated. To allow access to these calls there are also calls
+
+- `energy_forces` for potential energy and allocating forces
+- `energy_forces!` for potential energy and non-allocating forces
+- `energy_forces_virial` for potential energy, allocating forces and virial
+- `energy_forces_virial!` for potential energy, non-allocating forces and virial
+
+These all are generated automatically, if you have defined the corresponding individual methods. The main idea here is that you can implement more efficient methods by your self.
+
+Example implementation
+
+```julia
+function AtomsCalculators.potential_energy_forces(system, calculator::MyType; kwargs...)
+    # we can ignore kwargs... or use them to tune the calculation
+    # or give extra information like pairlist
+
+    # add your own definition here
+    E = 0.0u"eV"
+    f = zeros(AtomsCalculators.default_force_eltype, length(system))
+    return (;
```

Comment:
Is this the same as a NamedTuple? If yes, then it's an odd way (for my taste) to write this. If it is not the same, then why not use a NamedTuple?

Reply:
Yes it is NamedTuple. This is just my preferred way of writing NamedTuples.